WooCommerce/Classes/ViewRelated/Dashboard/Factories/DashboardUIFactory.swift

```diff
@@ -75,16 +76,60 @@ final class DashboardUIFactory {
 }
 
 private extension DashboardUIFactory {
-    func onStatsVersionStateChange(state: StatsVersionState, onUIUpdate: @escaping (_ dashboardUI: DashboardUI) -> Void) {
-        switch state {
+    func onStatsVersionStateChange(previousState: StatsVersionState?,
+                                   currentState: StatsVersionState,
+                                   onUIUpdate: @escaping (_ dashboardUI: DashboardUI) -> Void) {
+        switch currentState {
         case .initial(let statsVersion), .eligible(let statsVersion):
-            onUIUpdate(dashboardUI(siteID: siteID, statsVersion: statsVersion))
+            let updatedDashboardUI = dashboardUI(siteID: siteID, statsVersion: statsVersion)
+            onUIUpdate(updatedDashboardUI)
+
+            if let topBannerPresenter = updatedDashboardUI as? TopBannerPresenter {
+                topBannerPresenter.hideTopBanner(animated: true)
+            }
         case .v3ShownV4Eligible:
-            // TODO-1232: handle v3 --> v4 upgrading: shows top banner to announce stats v4 feature for user to opt in.
-            onUIUpdate(dashboardUI(siteID: siteID, statsVersion: .v4))
+            let updatedDashboardUI = dashboardUI(siteID: siteID, statsVersion: .v3)
+            onUIUpdate(updatedDashboardUI)
+
+            guard previousState != currentState else {
+                return
+            }
+
+            guard let topBannerPresenter = updatedDashboardUI as? TopBannerPresenter else {
```

Comment:
And here is the TopBannerPresenter added in the previous PR

Reply:
Silly question, but if we fail and assertion when updateDashBoardUI returns something that is not a TopBannerPresenter, why not make updateDashBoardUI return either something typed as TopBannerPresenter, or DashBoardUI & TopBannerPresenter?

Reply:
the tricky part is only stats v3 dashboard UI conforms to `TopBannerPresenter` - how about moving the guard check for `DashboardUI & TopBannerPresenter` earlier in the cases that require a `TopBannerPresenter` dashboard UI in https://github.com/woocommerce/woocommerce-ios/pull/1247/commits/2fb94293482c0f37c25f2cf91eca5924992cd25e ?

alternatively, I can also make stats v3 dashboard UI a separate function that returns `DashboardUI & TopBannerPresenter`. do you think this would be clearer than the guard statement?

Reply:
> alternatively, I can also make stats v3 dashboard UI a separate function that returns DashboardUI & TopBannerPresenter. do you think this would be clearer than the guard statement?

I think that would be better. Not because of clarity only, but because, if I understand correctly, now we'd have a method that would not return just one type, but at least two.

Reply:
just updated in https://github.com/woocommerce/woocommerce-ios/pull/1247/commits/d8ba75f317f5c461767d9f6c6ea1723c4992269f for another look!